lib/Watcher.js

```diff
@@ -0,0 +1,184 @@
+    onFilesAdded = (): Promise<void> => Promise.resolve(),
+    onFilesRemoved = (): Promise<void> => Promise.resolve(),
+    storage,
+  }: Object) {
+    this.workingDirectory = workingDirectory;
+    this.excludes = excludes;
+    this.onFilesAdded = onFilesAdded;
+    this.onFilesRemoved = onFilesRemoved;
+    this.storage = storage;
+  }
+
+  subscribe({
+    client,
+    fbWatch,
+    relativePath,
+  }: {
+    client: fbWatchman.Client,
+    fbWatch: string,
+    relativePath: string,
+  }): Promise<void> {
+    const subscription = {
+      // Match javascript files
+      expression: [
+        'anyof',
+        ['suffix', 'js'],
+        ['suffix', 'jsx'],
+        ['suffix', 'json'],
+      ],
+      fields: ['name', 'exists', 'mtime_ms'],
+      relative_root: relativePath,
+    };
+
+    return new Promise((resolve: Function, reject: Function) => {
+      client.command(['subscribe', fbWatch, SUBSCRIPTION_NAME, subscription],
+        (error: Error) => {
+          if (error) {
+            reject(error);
+            return;
+          }
+          resolve();
+        });
+
+      client.on('subscription', (resp: Object) => {
+        if (resp.subscription !== SUBSCRIPTION_NAME) {
+          return;
+        }
+
+        const added = [];
+        const removed = [];
+        resp.files.forEach((file: Object) => {
+          const normalizedPath = normalizePath(file.name, this.workingDirectory);
+          if (normalizedPath.indexOf('/node_modules/') !== -1) {
+            return;
+          }
+          if (this.excludes.some((pattern: string): boolean =>
+            minimatch(normalizedPath, pattern))) {
+            return;
+          }
+          if (file.exists) {
+            added.push({ path: normalizedPath, mtime: +file.mtime_ms });
+          } else {
+            removed.push({ path: normalizedPath });
+          }
+        });
+        if (added.length) {
+          this.onFilesAdded(added);
+        }
+        if (removed.length) {
+          this.onFilesRemoved(removed);
+        }
+      });
+    });
+  }
+
+  startSubscription({ client }: { client: fbWatchman.Client }): Promise<void> {
+    return new Promise((resolve: Function, reject: Function) => {
+      client.command(['watch-project', this.workingDirectory], (error: Error, resp: Object) => {
+        if (error) {
+          reject(error);
+          return;
+        }
+
+        if ('warning' in resp) {
+          loglevel.warn(`WARNING received during watchman init: ${resp.warning}`);
+        }
+
+        this.subscribe({
+          client,
+          fbWatch: resp.watch,
+          relativePath: resp.relative_path,
+        }).then(resolve).catch(reject);
+      });
+    });
+  }
+
+  initialize(): Promise<void> {
+    return new Promise((resolve: Function, reject: Function) => {
+      this.initializeWatchman().then(resolve).catch(() => {
+        loglevel.info(
+          "Couldn't initialize watchman watcher. Falling back to polling.");
```

Comment:
This might happen elsewhere, but it might be nice to provide more information about why unable to init watchman.